File: 10-05-2024/Combination Sum II.js
//{ Driver Code Starts

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => { inputString += inputStdin; });

process.stdin.on('end', _ => {
    inputString =
        inputString.trim().split("\n").map(string => { return string.trim(); });

    main();
});

function readLine() { return inputString[currentLine++]; }

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for (; i < t; i++) {

        let p = readLine().split(' ').map(x => parseInt(x));
        let n = p[0];
        let k = p[1];

        let arr = new Array(n);
        let input_arr = readLine().split(' ').map(x => parseInt(x));
        for (let i = 0; i < n; i++) {
            arr[i] = input_arr[i];
        }

        let obj = new Solution();
        let res = obj.CombinationSum2(arr, n, k);

        let S_res = '';
        for (let row of res) {
            let tmp = '';
            for (let col of row) {
                tmp += col;
                tmp += ' ';
            }
            console.log(tmp);
        }
    }
}

// } Driver Code Ends



class Solution {
    /**
     * @param number n
     * @param number[] arr
     * @param number k
     *
     * @returns number[][]
     */
    CombinationSum2(arr, n, k) {
        // Your Code Start
            // Sort the input array in ascending order
        arr.sort((a, b) => a - b);
        const result = []; // Initialize an array to store the combinations
        
        // Define a backtracking function to find combinations
        const backtrack = (start, target, curr) => {
            // If the target sum is reached, add the combination to the result
            if (target === 0) {
                result.push([...curr]);
                return;
            }
            // If target becomes negative, or all elements are used, or current element exceeds target, return
            if (target < 0 || start === arr.length || target < arr[start]) {
                return;
            }
            // Iterate through the array starting from 'start'
            for (let i = start; i < arr.length; i++) {
                // Skip duplicates
                if (i > start && arr[i] === arr[i - 1]) {
                    continue;
                }
                // Add current element to the combination
                curr.push(arr[i]);
                // Recursively call backtrack with updated parameters
                backtrack(i + 1, target - arr[i], curr);
                // Backtrack: Remove the last added element from the combination
                curr.pop();
            }
        };
        
        // Start backtracking from index 0 with initial target as 'k' and empty combination
        backtrack(0, k, []);
        // If no valid combinations found, return an empty array, otherwise return the result
        return result.length === 0 ? [[]] : result;
      //Your Code End
    }
}

File: 10-05-2024/Combination Sum II.cs
//{ Driver Code Starts
// Initial Template for C#

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DriverCode {

class GFG {
    static void Main(string[] args) {
        int testcases; // Taking testcase as input
        testcases = Convert.ToInt32(Console.ReadLine());
        while (testcases-- > 0) // Looping through all testcases
        {

            var ip = Console.ReadLine().Trim().Split(' ');
            int n = int.Parse(ip[0]);
            int k = int.Parse(ip[1]);
            List<int> arr = new List<int>();
            ip = Console.ReadLine().Trim().Split(' ');
            for (int i = 0; i < n; i++) {
                arr.Add(int.Parse(ip[i]));
            }
            Solution obj = new Solution();
            var res = obj.CombinationSum2(arr, n, k);
            foreach (var v in res) {
                foreach (var vv in v) {
                    Console.Write(vv + " ");
                }
                Console.WriteLine();
            }
            if (res.Count == 0) Console.WriteLine();
        }
    }
}
}
// } Driver Code Ends


// User function Template for C#


class Solution {
    // Complete this function
    public List<List<int>> CombinationSum2(List<int> arr, int n, int k) {
        List<List<int>> result = new List<List<int>>();
        arr.Sort(); // Sort the array to handle duplicates
        
        // Call the recursive helper function
        FindCombinations(arr, 0, new List<int>(), k, result);
        
        return result;
    }
    
    private void FindCombinations(List<int> arr, int index, List<int> current, int target, List<List<int>> result) {
        if (target == 0) {
            result.Add(new List<int>(current));
            return;
        }
        
        for (int i = index; i < arr.Count; i++) {
            // Skip duplicates
            if (i > index && arr[i] == arr[i - 1]) continue;
            
            // If current element exceeds target, break
            if (arr[i] > target) break;
            
            // Add current element to the combination
            current.Add(arr[i]);
            
            // Recur with reduced target and next index
            FindCombinations(arr, i + 1, current, target - arr[i], result);
            
            // Backtrack - remove the last element to try other combinations
            current.RemoveAt(current.Count - 1);
        }
    }
}
